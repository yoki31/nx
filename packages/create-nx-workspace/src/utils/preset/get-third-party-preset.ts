import { validateNpmPackage } from '../validate-npm-package';
import { isKnownPreset } from './preset';

/**
 * This function is used to check if a preset is a third party preset.
 * @param preset
 * @returns
 * - undefined if the preset is a known Nx preset or invalid.
 * - packageName if the preset is a third party preset.
 * - throws an error if the preset is invalid.
 */
export function getPackageNameFromThirdPartyPreset(
  preset?: string
): string | undefined {
  if (!preset || isKnownPreset(preset)) {
    return;
  }
  // extract the package name from the preset
  const packageName = preset.match(/.+@/)
    ? preset[0] + preset.substring(1).split('@')[0]
    : preset;

  const validateResult = validateNpmPackage(packageName);
  if (!validateResult.validForNewPackages) {
    throw new Error(
      `Invalid preset npm package ${packageName}. There was an error with the preset npm package you provided: ` +
        (validateResult.errors ?? []).join('\n')
    );
  }
  return packageName;
}
