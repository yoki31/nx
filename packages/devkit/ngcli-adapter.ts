/**
 * @category Ng CLI Adapter
 */
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
export {
  wrapAngularDevkitSchematic,
  mockSchematicsForTesting,
  NxScopedHost,
} from 'nx/src/adapter/ngcli-adapter';
