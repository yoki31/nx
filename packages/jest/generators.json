{
  "name": "Nx Jest",
  "version": "0.1",
  "generators": {
    "init": {
      "factory": "./src/generators/init/init#jestInitGeneratorInternal",
      "schema": "./src/generators/init/schema.json",
      "description": "Initialize the `@nx/jest` plugin.",
      "aliases": ["ng-add"],
      "hidden": true
    },
    "configuration": {
      "factory": "./src/generators/configuration/configuration#configurationGeneratorInternal",
      "schema": "./src/generators/configuration/schema.json",
      "description": "Add Jest configuration to a project.",
      "hidden": true
    },
    "convert-to-inferred": {
      "factory": "./src/generators/convert-to-inferred/convert-to-inferred",
      "schema": "./src/generators/convert-to-inferred/schema.json",
      "description": "Convert existing Jest project(s) using `@nx/jest:jest` executor to use `@nx/jest/plugin`."
    }
  }
}
