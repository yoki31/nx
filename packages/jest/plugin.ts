export {
  createNodes,
  createNodesV2,
  JestPluginOptions,
} from './src/plugins/plugin';
