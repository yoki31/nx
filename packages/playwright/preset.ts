export * from './src/utils/preset';
