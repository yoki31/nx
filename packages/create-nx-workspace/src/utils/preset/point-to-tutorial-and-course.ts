import { output } from '../output';
import { Preset } from './preset';

export function pointToTutorialAndCourse(preset: Preset) {
  const title = `First time using Nx? Check out this interactive Nx tutorial.`;
  switch (preset) {
    case Preset.NPM:
    case Preset.Apps:
      output.addVerticalSeparator();
      output.note({
        title,
        bodyLines: [
          `https://nx.dev/getting-started/tutorials/npm-workspaces-tutorial`,
        ],
      });
      break;

    case Preset.TS:
      output.addVerticalSeparator();
      output.note({
        title,
        bodyLines: [
          `https://nx.dev/getting-started/tutorials/integrated-repo-tutorial`,
        ],
      });
      break;
    case Preset.ReactStandalone:
      output.addVerticalSeparator();
      output.note({
        title,
        bodyLines: [
          `https://nx.dev/getting-started/tutorials/react-standalone-tutorial`,
        ],
      });
      break;
    case Preset.ReactMonorepo:
    case Preset.NextJs:
    case Preset.NextJsStandalone:
      output.addVerticalSeparator();
      output.note({
        title,
        bodyLines: [`https://nx.dev/react-tutorial/1-code-generation`],
      });
      break;
    case Preset.AngularStandalone:
      output.addVerticalSeparator();
      output.note({
        title,
        bodyLines: [
          `https://nx.dev/getting-started/tutorials/angular-standalone-tutorial`,
        ],
      });
      break;
    case Preset.AngularMonorepo:
      output.addVerticalSeparator();
      output.note({
        title,
        bodyLines: [`https://nx.dev/angular-tutorial/1-code-generation`],
      });
      break;
    case Preset.Express:
    case Preset.NodeStandalone:
      output.addVerticalSeparator();
      output.note({
        title,
        bodyLines: [
          `https://nx.dev/getting-started/tutorials/node-server-tutorial`,
        ],
      });
      break;
  }
}
