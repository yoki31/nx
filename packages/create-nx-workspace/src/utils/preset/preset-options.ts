import { Preset } from './preset';

export const presetOptions: { name: Preset; message: string }[] = [
  {
    name: Preset.Apps,
    message:
      'apps              [an empty monorepo with no plugins with a layout that works best for building apps]',
  },
  {
    name: Preset.TS,
    message:
      'ts                [an empty monorepo with the JS/TS plugin preinstalled]',
  },
  {
    name: Preset.ReactMonorepo,
    message: 'react             [a monorepo with a single React application]',
  },
  {
    name: Preset.AngularMonorepo,
    message: 'angular           [a monorepo with a single Angular application]',
  },
  {
    name: Preset.VueMonorepo,
    message: 'vue               [a monorepo with a single Vue application]',
  },
  {
    name: Preset.Nuxt,
    message: 'nuxt              [a monorepo with a single Nuxt application]',
  },
  {
    name: Preset.NextJs,
    message: 'next.js           [a monorepo with a single Next.js application]',
  },
  {
    name: Preset.Nest,
    message: 'nest              [a monorepo with a single Nest application]',
  },
  {
    name: Preset.ReactNative,
    message:
      'react-native      [a monorepo with a single React Native application]',
  },
];
