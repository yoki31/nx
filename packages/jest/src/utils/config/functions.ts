import type * as ts from 'typescript';
import { applyChangesToString, ChangeType, Tree } from '@nx/devkit';
import { Config } from '@jest/types';
import { createContext, runInContext } from 'vm';
import { dirname, join } from 'path';
import { ensureTypescript } from '@nx/js/src/utils/typescript/ensure-typescript';

let tsModule: typeof import('typescript');

function makeTextToInsert(
  value: unknown,
  precedingCommaNeeded: boolean
): string {
  return `${precedingCommaNeeded ? ',' : ''}${value}`;
}

function findPropertyAssignment(
  object: ts.ObjectLiteralExpression,
  propertyName: string
) {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }

  return object.properties.find((prop) => {
    if (!tsModule.isPropertyAssignment(prop)) {
      return false;
    }
    const propNameText = prop.name.getText();
    if (propNameText.match(/^["'].+["']$/g)) {
      return JSON.parse(propNameText.replace(/'/g, '"')) === propertyName;
    }

    return propNameText === propertyName;
  }) as ts.PropertyAssignment | undefined;
}

export function addOrUpdateProperty(
  tree: Tree,
  object: ts.ObjectLiteralExpression,
  properties: string[],
  value: unknown,
  path: string
) {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }
  const { SyntaxKind } = tsModule;

  const propertyName = properties.shift();
  const propertyAssignment = findPropertyAssignment(object, propertyName);

  const originalContents = tree.read(path, 'utf-8');

  if (propertyAssignment) {
    if (
      propertyAssignment.initializer.kind === SyntaxKind.StringLiteral ||
      propertyAssignment.initializer.kind === SyntaxKind.NumericLiteral ||
      propertyAssignment.initializer.kind === SyntaxKind.FalseKeyword ||
      propertyAssignment.initializer.kind === SyntaxKind.TrueKeyword
    ) {
      const updatedContents = applyChangesToString(originalContents, [
        {
          type: ChangeType.Delete,
          start: propertyAssignment.initializer.pos,
          length: propertyAssignment.initializer.getFullText().length,
        },
        {
          type: ChangeType.Insert,
          index: propertyAssignment.initializer.pos,
          text: value as string,
        },
      ]);

      tree.write(path, updatedContents);
      return;
    }

    if (
      propertyAssignment.initializer.kind === SyntaxKind.ArrayLiteralExpression
    ) {
      const arrayLiteral =
        propertyAssignment.initializer as ts.ArrayLiteralExpression;

      if (
        arrayLiteral.elements.some((element) => {
          return element.getText().replace(/'/g, '"') === value;
        })
      ) {
        return [];
      }

      if (arrayLiteral.elements.length === 0) {
        const updatedContents = applyChangesToString(originalContents, [
          {
            type: ChangeType.Insert,
            index: arrayLiteral.elements.end,
            text: value as string,
          },
        ]);
        tree.write(path, updatedContents);
        return;
      } else {
        const text = makeTextToInsert(
          value,
          arrayLiteral.elements.length !== 0 &&
            !arrayLiteral.elements.hasTrailingComma
        );
        const updatedContents = applyChangesToString(originalContents, [
          {
            type: ChangeType.Insert,
            index: arrayLiteral.elements.end,
            text,
          },
        ]);
        tree.write(path, updatedContents);
        return;
      }
    } else if (
      propertyAssignment.initializer.kind === SyntaxKind.ObjectLiteralExpression
    ) {
      return addOrUpdateProperty(
        tree,
        propertyAssignment.initializer as ts.ObjectLiteralExpression,
        properties,
        value,
        path
      );
    }
  } else {
    if (propertyName === undefined) {
      throw new Error(
        `Please use dot delimited paths to update an existing object. Eg. object.property `
      );
    }
    const text = makeTextToInsert(
      `${JSON.stringify(propertyName)}: ${value}`,
      object.properties.length !== 0 && !object.properties.hasTrailingComma
    );
    const updatedContents = applyChangesToString(originalContents, [
      {
        type: ChangeType.Insert,
        index: object.properties.end,
        text,
      },
    ]);
    tree.write(path, updatedContents);
    return;
  }
}

export function removeProperty(
  object: ts.ObjectLiteralExpression,
  properties: string[]
): ts.PropertyAssignment | null {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }

  const propertyName = properties.shift();
  const propertyAssignment = findPropertyAssignment(object, propertyName);

  if (propertyAssignment) {
    if (
      properties.length > 0 &&
      propertyAssignment.initializer.kind ===
        tsModule.SyntaxKind.ObjectLiteralExpression
    ) {
      return removeProperty(
        propertyAssignment.initializer as ts.ObjectLiteralExpression,
        properties
      );
    }
    return propertyAssignment;
  } else {
    return null;
  }
}

function isModuleExport(node: ts.Statement) {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }

  return (
    tsModule.isExpressionStatement(node) &&
    node.expression?.kind &&
    tsModule.isBinaryExpression(node.expression) &&
    node.expression.left.getText() === 'module.exports' &&
    node.expression.operatorToken?.kind === tsModule.SyntaxKind.EqualsToken
  );
}

function isDefaultExport(node: ts.Statement) {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }

  return (
    tsModule.isExportAssignment(node) &&
    node.expression?.kind &&
    tsModule.isObjectLiteralExpression(node.expression) &&
    node.getText().startsWith('export default')
  );
}

/**
 * Should be used to get the jest config object as AST
 */
export function jestConfigObjectAst(
  fileContent: string
): ts.ObjectLiteralExpression {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }

  const sourceFile = tsModule.createSourceFile(
    'jest.config.ts',
    fileContent,
    tsModule.ScriptTarget.Latest,
    true
  );

  const exportStatement = sourceFile.statements.find(
    (statement) => isModuleExport(statement) || isDefaultExport(statement)
  );

  let ast: ts.ObjectLiteralExpression;
  if (tsModule.isExpressionStatement(exportStatement)) {
    const moduleExports = exportStatement.expression as ts.BinaryExpression;
    if (!moduleExports) {
      throw new Error(
        `
       The provided jest config file does not have the expected 'module.exports' expression. 
       See https://jestjs.io/docs/en/configuration for more details.`
      );
    }

    ast = moduleExports.right as ts.ObjectLiteralExpression;
  } else if (tsModule.isExportAssignment(exportStatement)) {
    const defaultExport =
      exportStatement.expression as ts.ObjectLiteralExpression;

    if (!defaultExport) {
      throw new Error(
        `
       The provided jest config file does not have the expected 'export default' expression. 
       See https://jestjs.io/docs/en/configuration for more details.`
      );
    }

    ast = defaultExport;
  }
  if (!ast) {
    throw new Error(
      `
      The provided jest config file does not have the expected 'module.exports' or 'export default' expression. 
      See https://jestjs.io/docs/en/configuration for more details.`
    );
  }

  if (!tsModule.isObjectLiteralExpression(ast)) {
    throw new Error(
      `The 'export default' or 'module.exports' expression is not an object literal.`
    );
  }

  return ast;
}

/**
 * Returns the jest config object
 * @param host
 * @param path
 */
export function jestConfigObject(
  host: Tree,
  path: string
): Partial<Config.InitialOptions> & { [index: string]: any } {
  const __filename = join(host.root, path);
  const contents = host.read(path, 'utf-8');
  let module = { exports: {} };

  // transform the export default syntax to module.exports
  // this will work for the default config, but will break if there are any other ts syntax
  // TODO(caleb): use the AST to transform back to the module.exports syntax so this will keep working
  //  or deprecate and make a new method for getting the jest config object
  const forcedModuleSyntax = contents.replace(
    /export\s+default/,
    'module.exports ='
  );
  // Run the contents of the file with some stuff from this current context
  // The module.exports will be mutated by the contents of the file...
  runInContext(
    forcedModuleSyntax,
    createContext({
      module,
      require,
      process,
      __filename,
      __dirname: dirname(__filename),
    })
  );

  // TODO: jest actually allows defining configs with async functions... we should be able to read that...
  return module.exports;
}
