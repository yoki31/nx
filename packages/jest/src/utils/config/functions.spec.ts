import { createTree } from '@nx/devkit/testing';
import { jestConfigObject, jestConfigObjectAst } from './functions';

describe('jestConfigObject', () => {
  describe('module.exports', () => {
    it('should work for basic cases', () => {
      const tree = createTree();
      tree.write(
        'jest.config.js',
        `
      module.exports = {
        foo: 'bar'
      };
    `
      );

      expect(jestConfigObject(tree, 'jest.config.js')).toEqual({
        foo: 'bar',
      });
    });

    xit('should work with async functions', async () => {
      const tree = createTree();
      jest.mock('@nx/jest', () => ({
        getJestProjects: () => ['<rootDir>/project-a', '<rootDir>/project-b'],
      }));
      tree.write(
        'jest.config.js',
        `
      const { getJestProjects } = require('@nx/jest');
      module.exports = async () => ({
        foo: 'bar'
      });
    `
      );

      expect(await jestConfigObject(tree, 'jest.config.js')).toEqual({
        foo: 'bar',
      });
    });

    it('should work with `getJestConfig`', () => {
      const tree = createTree();
      jest.mock('@nx/jest', () => ({
        getJestProjects: () => ['<rootDir>/project-a', '<rootDir>/project-b'],
      }));
      tree.write(
        'jest.config.js',
        `
      const { getJestProjects } = require('@nx/jest');
      module.exports = {
        projects: getJestProjects()
      };
    `
      );

      expect(jestConfigObject(tree, 'jest.config.js')).toEqual({
        projects: ['<rootDir>/project-a', '<rootDir>/project-b'],
      });
    });

    it('should work with node globals (require, __dirname, process, __filename, console, and other globals)', () => {
      const tree = createTree();
      jest.mock('@nx/jest', () => ({
        getJestProjects: () => ['<rootDir>/project-a', '<rootDir>/project-b'],
      }));
      tree.write(
        'jest.config.js',
        `
      const { getJestProjects } = require('@nx/jest');
      module.exports = {
        projects: getJestProjects(),
        filename: __filename,
        env: process.env,
        dirname: __dirname
      };
    `
      );

      expect(jestConfigObject(tree, 'jest.config.js')).toEqual({
        dirname: '/virtual',
        filename: '/virtual/jest.config.js',
        env: process.env,
        projects: ['<rootDir>/project-a', '<rootDir>/project-b'],
      });
    });
  });

  describe('export default', () => {
    it('should work for basic cases', () => {
      const content = `
        export default {
          abc: 'xyz'
        }`;

      expect(jestConfigObjectAst(content).getText()).toMatchSnapshot();
    });

    it('should handle spread assignments', () => {
      const content = `
       import { nxPreset } from '@nx/jest/preset';
      
        export default {
          ...nxPreset,
          abc: 'xyz'
        }`;

      expect(jestConfigObjectAst(content).getText()).toMatchSnapshot();
    });
  });
});
