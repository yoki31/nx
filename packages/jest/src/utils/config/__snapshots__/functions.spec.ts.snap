// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jestConfigObject export default should handle spread assignments 1`] = `
"{
          ...nxPreset,
          abc: 'xyz'
        }"
`;

exports[`jestConfigObject export default should work for basic cases 1`] = `
"{
          abc: 'xyz'
        }"
`;
